File: test/jvm/backtype/storm/utils/UtilsTest.java
package backtype.storm.utils;

import org.junit.Test;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import com.netflix.curator.CuratorZookeeperClient;
import com.netflix.curator.framework.CuratorFramework;
import com.netflix.curator.retry.ExponentialBackoffRetry;

import backtype.storm.Config;

public class UtilsTest {

  @Test
  public void testNewCuratorUsesBoundedExponentialBackoff() {
    @SuppressWarnings("unchecked")
    Map<String,Object> conf = (Map<String,Object>)Utils.readDefaultConfig();

    // Ensure these two values are different.
    final int ArbitraryInterval = 24;
    final int ArbitraryRetries = 4;
    conf.put(Config.STORM_ZOOKEEPER_RETRY_INTERVAL, ArbitraryInterval);
    conf.put(Config.STORM_ZOOKEEPER_RETRY_TIMES, ArbitraryRetries);

    List<String> servers = new ArrayList<String>();
    servers.add("bogus_server");
    Object port = new Integer(42);
    CuratorFramework cf = Utils.newCurator(conf, servers, port);

    assertTrue(cf.getZookeeperClient().getRetryPolicy() 
            instanceof ExponentialBackoffRetry);

    ExponentialBackoffRetry retry =
        (ExponentialBackoffRetry)cf.getZookeeperClient().getRetryPolicy();
    assertEquals(retry.getBaseSleepTimeMs(), ArbitraryInterval);
    assertEquals(retry.getN(), ArbitraryRetries);
  }
}

File: test/clj/backtype/storm/utils_test.clj
(ns backtype.storm.utils-test
  (:import [backtype.storm Config])
  (:import [backtype.storm.utils Utils])
  (:import [com.netflix.curator.retry ExponentialBackoffRetry])
  (:use [backtype.storm util])
  (:use [clojure test])
)

(deftest test-new-curator-uses-exponential-backoff
  (let [expected_interval 2400
        expected_retries 10
        expected_ceiling (/ expected_interval 2)
        conf (merge (clojurify-structure (Utils/readDefaultConfig))
          {Config/STORM_ZOOKEEPER_RETRY_INTERVAL expected_interval
           Config/STORM_ZOOKEEPER_RETRY_TIMES expected_retries
           Config/STORM_ZOOKEEPER_RETRY_INTERVAL_CEILING expected_ceiling})
        servers ["bogus_server"]
        arbitrary_port 42
        curator (Utils/newCurator conf servers arbitrary_port)
        retry (-> curator .getZookeeperClient .getRetryPolicy)
       ]
    (is (.isAssignableFrom ExponentialBackoffRetry (.getClass retry)))
    (is (= (.getBaseSleepTimeMs retry) expected_interval)) 
    (is (= (.getN retry) expected_retries)) 
    (is (= (.getMaxRetryInterval retry) expected_ceiling)) 
    (is (= (.getSleepTimeMs retry 10 0) expected_ceiling))
  )
)
